disperser/encoder/server_v2.go

```diff
@@ -0,0 +1,222 @@
+package encoder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/common/healthcheck"
+	v2 "github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigenda/disperser"
+	pb "github.com/Layr-Labs/eigenda/disperser/api/grpc/encoder/v2"
+	"github.com/Layr-Labs/eigenda/disperser/common/v2/blobstore"
+	"github.com/Layr-Labs/eigenda/encoding"
+	"github.com/Layr-Labs/eigenda/encoding/rs"
+	"github.com/Layr-Labs/eigenda/relay/chunkstore"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+type EncoderServerV2 struct {
+	pb.UnimplementedEncoderServer
+
+	config      ServerConfig
+	blobStore   *blobstore.BlobStore
+	chunkWriter chunkstore.ChunkWriter
+	logger      logging.Logger
+	prover      encoding.Prover
+	metrics     *Metrics
+	close       func()
+
+	runningRequests chan struct{}
+	requestPool     chan struct{}
+}
+
+func NewEncoderServerV2(config ServerConfig, blobStore *blobstore.BlobStore, chunkWriter chunkstore.ChunkWriter, logger logging.Logger, prover encoding.Prover, metrics *Metrics) *EncoderServerV2 {
+	return &EncoderServerV2{
+		config:      config,
+		blobStore:   blobStore,
+		chunkWriter: chunkWriter,
+		logger:      logger.With("component", "EncoderServer"),
+		prover:      prover,
+		metrics:     metrics,
+
+		runningRequests: make(chan struct{}, config.MaxConcurrentRequests),
+		requestPool:     make(chan struct{}, config.RequestPoolSize),
+	}
+}
+
+func (s *EncoderServerV2) Start() error {
+	// Serve grpc requests
+	addr := fmt.Sprintf("%s:%s", disperser.Localhost, s.config.GrpcPort)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("Could not start tcp listener: %v", err)
+	}
+
+	gs := grpc.NewServer()
+	reflection.Register(gs)
+	pb.RegisterEncoderServer(gs, s)
+
+	// Register Server for Health Checks
+	name := pb.Encoder_ServiceDesc.ServiceName
+	healthcheck.RegisterHealthServer(name, gs)
+
+	s.close = func() {
+		err := listener.Close()
+		if err != nil {
+			log.Printf("failed to close listener: %v", err)
+		}
+		gs.GracefulStop()
+	}
+
+	s.logger.Info("port", s.config.GrpcPort, "address", listener.Addr().String(), "GRPC Listening")
+	return gs.Serve(listener)
+}
+
+func (s *EncoderServerV2) EncodeBlobToChunkStore(ctx context.Context, req *pb.EncodeBlobRequest) (*pb.EncodeBlobToChunkStoreReply, error) {
+	totalStart := time.Now()
+	defer func() {
+		s.metrics.ObserveLatency("total", time.Since(totalStart))
+	}()
+
+	// Rate limit
+	select {
+	case s.requestPool <- struct{}{}:
```

Comment:
do you think this is still a good way to rate limit requests?

Reply:
I'll discuss with @jianoaix when he gets back but leaving it as is for now. I think we wanted to add queueing instead of immediately rejecting a request if the pool is full and use that as backpressure signal.

Reply:
Sg to keep it for now.